content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,324 @@
+---
+title: "How to use CompletableFuture to improve Performance"
+categories: ["Java"]
+date: 2022-02-08T11:00:00 
+authors: ["nukajbl"]
+excerpt: "In this article, we will see how to increase the performance of our application using the class
+CompletableFuture."
+image: images/stock/0117-future-1200x628-branded.jpg 
+url: api-calls-performance-with-completable-future
+---
+
+In this article, we will learn how to use `CompletableFuture` to increase the performance of our application. In the
+beginning, we will give the definition of the `Future` interface and what are its limitations and later will discuss how
+we can instead use the `CompletableFuture` class to overcome these limitations. We will do this by building a simple
+application that tries to categorize a list of bank `Transactions` using an external service, in our case we will just
+mock it and add some delay to simulate a remote call. Let's begin our journey!
+
+## What Is A `Future`?
+
+`Future` is a Java interface that was introduced in Java 5 to represent a value that will be available in the future.
+The advantages of using a `Future` are enormous because we could do some very intensive computation asynchronously
+without blocking the current thread that in the meantime can do some other useful job.
+
+We can think of it as going to the restaurant. During the time that the chef is preparing our dinner, we can do other
+things, like talking to friends or drinking a glass of wine and once the chef has finished the preparation, we can
+finally eat. Another advantage is that using the `Future` interface is much more developer-friendly than working
+directly with threads.
+
+## Limitations Of `Future`
+
+- We cannot complete manually the operation by providing a value  
+  Let's imagine that in our case, the categorization service that we are using to get the category for our transactions,
+  goes down, we would like to complete the  `Future` manually by providing a default value. This is not possible using
+  the `Future` interface.
+- We cannot combine one or more asynchronous operations  
+  Using the same example of categorization service, we would like to call another API passing the value of the category
+  returned by our categorization service, creating a pipeline. But this scenario is not possible using the `Future`.
+- There is no way to react to the completion of the asynchronous operation without blocking the thread by
+  invoking [get](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get())
+  method of the `Future` interface
```

Comment:
These limitations are hard to grasp without code examples. I suggest to move the limitations into the next section and then make a sub-section for each limitation with a code example that shows the limitation with `Future` and then another code example that shows how you can do it better with `CompletableFuture`.

I.e.:

* `CompletableFuture` vs. `Future`
  * Manually Completing an operation
  * Combining Asynchronous Operations
  * Blocking

and each section would contain the limitation with `Future` and then the improvement with `CompletableFuture`. That is much easier to read than listing the limitations up front and then discussing them again later.

Reply:
So I can remove this section and let only **CompletableFuture vs. Future** ? I didn't include code for these limitations for Future because I was thinking I can not write code for that because those methods don't exist for `Future`

Reply:
@thombergs pls answer to my question above

Reply:
Sorry, I just saw this because of your ping in Trello.

Yes, remove this section and make each limitation its own subsection under "CompletableFuture vs. Future". If you don't have code examples for the `Future` limitations, at least explain them in more detail so that readers understand the limitations. I didn't understand them from the 2-sentence explanations in this list. I think that code examples would help to understand them, even if they show workarounds with `Future`.

Reply:
done